rfc-drafts/stream.md

```diff
@@ -63,14 +63,20 @@ the current task to be re-awoken when the data is ready.
 ```rust
 // Defined in std::stream module
 pub trait Stream {
+    // Core items:
     type Item;
-    
     fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>>;
     
+    // Optional optimization hint, just like with iterators:
     #[inline]
     fn size_hint(&self) -> (usize, Option<usize>) {
         (0, None)
     }
+
+    // Convenience methods (covered later on in the RFC):
+    fn next(&mut self) -> Next<'_, Self>
+    where
+        Self: Unpin;
```

Comment:
can somebody (maybe @withoutboats or @cramertj) explain to me why `Self: Unpin` is here and why that's not a crushing limitation? I feel like I'm missing something super basic. I guess the answer is maybe that we expect to be operating with `Box<S>` where `S: Stream`, and `Box<S>` is always `Unpin`?

And in terms of why it is needed, I guess it is because: you can invoke `next` and that creates a `&mut` references to self, but once you've consumed the item, you want to be able to move `self` somewhere else .. I don't know. I can see *why* it requires `Unpin` (we are invoking `Pin::new` in the `next` method on `Next`), but I'm having a hard time with the "intuition" here.

I guess the alternative would be the `next` takes `Pin<&mut Self>` and not `&mut self`? Why would that be bad?

Anyway, I think we should have some text justifying it.

Reply:
Sure, the reason why its needed is pretty clear if you look at the definition of Next:

``rust
struct Next<'a, S: Stream> {
    stream: &'a mut S
}
```

Since `Stream::poll_next` takes a pinned reference, the next future pretty clearly needs `S` to be `Unpin` to safely construct a `Pin<&mut S>` from `&mut S`. The alternative would be for next to take `Pin<&mut S>`.

You're basically right about operating on `Box<S>`, but remember that we can also operate on a stack-pinned `S` a lot of the time, so it doesn't necessarily require an allocation.

I think the current approach is strictly better than requiring the method to be pinned, because that would just require pinning it even when the type is unpin, whereas the current situation requires pinning it only when the type is unpin.

Right now, no one cares because we have no `!Unpin` streams in practice. But once we have async generators, that's going to be a big annoyance with using them. The problem is even worse with generator/Iterator, because generators won't implement Iterator if they have to be pinned to run, instead only pinned references to them will. I'm not sure what we're going to do about this when generators exist, this was one of the big papercuts we hadn't looked into yet.

Reply:
Yeah, so the problem will be that if `async gen` just returns an `impl Stream` (and not, crucially, an `impl Stream + Unpin`), you'll have to explicitly do some pinning operation to turn them into a "usable" stream (i.e., one you can call `next` one?).

The same seems to apply to a `gen fn`, but as you say, the situation is almost worse.

It does make me wonder if `Stream` should have a `poll_next(&mut self)` method instead of `poll_next(Pin<&mut self>)`, and we have the expectation that a `gen fn` should require some kind of "explicit operation" to make a `Stream` (which would basically be pinning it). The main benefit of this would be that `Iterator` kind of requires that. Basically, the argument to me for "let's just stabilize `Stream`" has relied in part on the idea that all the problems we face for Stream, we also face for Iterator, but actually in this situation the two designs diverge slightly (with Iterator potentially having "another problem", depending on your POV).

I guess that the idea would be that a `gen fn` actually doesn't return an `Iterator` but rather an `impl IntoIterator`, and an `async gen fn` returns an `impl IntoStream`, where the "into" part of it does the pinning required to make `&mut self` reasonable.

This seems like an important detail, I'm glad we're digging into it a bit more.

Reply:
> Right now, no one cares because we have no `!Unpin` streams in practice.

To make sure I'm following, this is because people are hand-writing their streams, and basically here the `Pin`-machinery is just kind of boilerplate.

Reply:
Here is an example when removing the `Unpin` bounds and using unsafe: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=526ee8e3ebca18a4a8756bff3984a23a) 
(this is unsound because you can move the underlying stream after `Next` dropped)

Reply:
> To make sure I'm following, this is because people are hand-writing their streams, and basically here the Pin-machinery is just kind of boilerplate.

Exactly. Hand-written streams/futures that are not `Unpin` are very rare. All the Pin machinery exists on the assumption that some day we will generate streams from functions and these will not implement Unpin.

Reply:
> I guess that the idea would be that a gen fn actually doesn't return an Iterator but rather an impl IntoIterator, and an async gen fn returns an impl IntoStream, where the "into" part of it does the pinning required to make &mut self reasonable.

The only way to make a trait that does this is to box it, since you can't write a function that returns a stack pinned value.

Since the big use case for iterators (and probably for streams if we ever have a dedicated syntax) is to loop with `for`, which inherently "pins" them in place to the stack, we'll probably want something a bit more nuanced than them returning IntoIterator, so that for loops can somehow treat them as pinned without making an allocation (which in addition to its costs doesn't work at all in no-std no-alloc environments).

Reply:
OK, I've been turning this over. Bear with me.

I'm basically trying to justify a design where `Stream` requires a pinned self, even though `Iterator` does not, given that we want `Stream` and `Iterator` to work as analogously as possible. I guess my main concern is that `Iterator` doesn't require pinning in its core `next` method, so for a `gen fn` to interoperate with that ecosystem, there is going to have to be some kind of initial pinning step that converts its result into an iterator. This is going to be a bit tricky since, as you point out, you can't return a pinned value except by boxing, at least today. 

The point though is that the general shape will be `gen_fn().pin_somehow().adapter1().adapter2()`.

On the other hand, with streams, the core interface *is* pinned, and so the pinning occurs at the last moment, kind of right before you invoke `next`. Something like `async_gen_fn().adapter1().adapter2().pin_somehow()`.

"Pin at end" is probably the better setup, since it lets you build and return those adapters and then apply to pinning at the end. There are perhaps also efficiency gains from this setup, since maybe the adapters want to be pinned and can take advantage of that. So this is probably the argument in favor of the current design (and of course the analogy with futures).

I guess that implies that to have a `gen fn` that produces iterators will require *either* 

* pinning at the start, vs the end, which is asymmetric but maybe that's ok
* disallowing borrowing yields, also asymmetric
* some kind of `PinnedIterator` trait that can be "adapted" into an iterator (by pinning)
* some clever idea I've not had yet

Does that sound right?

Reply:
Just to make your third idea more concrete, for my own sake:

```rust
trait PinIterator {
  type Item;
  fn next(self: Pin<&mut Self>) -> Self::Item;
  // combinators can go here (duplicating Iterator for the most part)
}
impl<I: PinIterator, P: Deref<Target = I> + DerefMut> Iterator for Pin<P> {
  type Item = <I as PinIterator>::Item;
  fn next(&mut self) -> Self::Item { self.as_mut().next() }
}

// this would be nice.. but would lead to name resolution ambiguity for our combinators 😬 
default impl<T: Iterator> PinIterator for T { .. }
```

Reply:
> I'm basically trying to justify a design where Stream requires a pinned self, even though Iterator does not, given that we want Stream and Iterator to work as analogously as possible.

The pinning also applies to the design of AsyncRead/AsyncWrite, which currently use Pin even though there's no clear plan to make them implemented with generator type syntax. In essence the asyncification of a singature is currently understood as pinned reciever + context arg + return poll.

> There are perhaps also efficiency gains from this setup, since maybe the adapters want to be pinned and can take advantage of that.

Pinning also introduces an indirection which will probably impact the optimizability of these adapters if you have to pin first.

> disallowing borrowing yields, also asymmetric

We should actually pursue this idea, it might work! Remember that we initially had only non-self-referential generators before adding self-referential generators that don't implement Unpin. It was immediately clear that holding a self-reference while doing nonblocking IO was super essential, but I'm not as certain that's true for iteration. It's possible that non-async generators could disallow self-references and just implement Unpin. That would certainly make the situation with Iterator much easier.

It's also possible that this would be a really bad limitation that makes them almost useless. But we need a real investigation to find out which it is.

Reply:
With respect to generators that cannot borrow across a yield, I think it's not going to work. I imagine a lot of people are going to want to write code like this

```rust
fn consume(x: impl Iterator<Item = T>) { .. }

consume(gen || {
    for item in myvec.iter() {
        if something(item) { continue; }
        if let Some(foo) = something_else(item) { yield foo; }
    }
});
```

and that will require holding a borrow across a yield.

Reply:
> The pinning also applies to the design of AsyncRead/AsyncWrite, which currently use Pin even though there's no clear plan to make them implemented with generator type syntax. In essence the asyncification of a singature is currently understood as pinned reciever + context arg + return poll.

Yes. I'm trying to push a bit on this and make sure I understand *why* we think that is the correct approach, I guess. =)

Reply:
So: I don't think that this conversation has to block this PR from landing, and it probably shouldn't even block the RFC itself, but I stll feel like there is *some* missing text around the `next` method that at minimum explains *why* `Unpin` is needed and perhaps tries to cover some of the reasoning above about why `poll_next` should take a pinned argument. Maybe the right place for some of this text is in the coverage of generator syntax, since we can discuss some of the difficulties that will have to be resolved in making generator syntax (and async generator syntax) work.

Reply:
> With respect to generators that cannot borrow across a yield, I think it's not going to work. I imagine a lot of people are going to want to write code like this

See unless you write `gen move`, this is not a self-referential type at all, because it can capture `myvec` by reference. :) Users writing futures combinators *immediately* ran into the problem of them being best expressed by self-referential types (and ended up having to Arc<Mutex> all kinds of stuff to share ownership across yield points), but users writing iterator combintors never have that problem. I think this is worth considering more seriously.

Reply:
The key to the difference is that futures are ultimately passed to some executor API like `spawn`, which expects a static future; to achieve that the futures contain all the state they need, and references are internal to that state. Iterators are almost never required to be `'static` by the APIs that consume them.

However, it's still the case that it will be far more natural to end up in self-referential situations with a generator than it is with a manually implemented iterator, because you will be able to implicitly build that state using the member variables of a function. So the restriction might still limit users from creating code in a natural way. And it's a bit confusing for the restriction to exist only in non-async generators.

So we'll have to get more experience to figure it out.